File: apps/web/src/components/origin/geometry.ts
"use client";

// Client-side mirror of the server-side normalized geometry shape.
// Kept in sync with verolas_api/workflow/origin/geometry.py. The shape
// is the canonical contract between the floor_parse adapter and the
// architectural-review editor.
//
// Distances are metres. Y-positive points up (engineering convention);
// the canvas applies a Y-flip transform when rendering so the user
// sees plans the right way around.

export interface Point2D {
  x: number;
  y: number;
}

export interface Extents {
  min_x: number;
  min_y: number;
  max_x: number;
  max_y: number;
}

export type WallKind = "exterior" | "interior" | "shear" | "unknown";
export type OpeningKind = "door" | "window" | "opening";

export interface Wall {
  id: string;
  start: Point2D;
  end: Point2D;
  thickness_m: number;
  kind: WallKind;
}

export interface Opening {
  id: string;
  wall_id: string;
  center: Point2D;
  width_m: number;
  kind: OpeningKind;
}

export interface Slab {
  id: string;
  polygon: Point2D[];
}

export interface Column {
  id: string;
  center: Point2D;
  size_m: [number, number];
}

export interface Floor {
  key: string;
  name: string;
  elevation_m: number;
  extents: Extents;
  walls: Wall[];
  openings: Opening[];
  slabs: Slab[];
  columns: Column[];
  is_roof: boolean;
}

export interface Geometry {
  source_format: "dxf" | "ifc";
  floors: Floor[];
  parser_notes: string[];
}

// Recompute a floor's bounding extents from its current entities.
// Edits change the geometry, so the original extents go stale; the
// canvas viewport reads the recomputed extents.
export function recomputeExtents(floor: Floor): Extents {
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  const track = (x: number, y: number): void => {
    if (x < minX) minX = x;
    if (y < minY) minY = y;
    if (x > maxX) maxX = x;
    if (y > maxY) maxY = y;
  };
  floor.walls.forEach((w) => {
    track(w.start.x, w.start.y);
    track(w.end.x, w.end.y);
  });
  floor.columns.forEach((c) => track(c.center.x, c.center.y));
  floor.openings.forEach((o) => track(o.center.x, o.center.y));
  floor.slabs.forEach((s) => s.polygon.forEach((p) => track(p.x, p.y)));
  if (minX === Infinity) {
    return { min_x: 0, min_y: 0, max_x: 0, max_y: 0 };
  }
  return { min_x: minX, min_y: minY, max_x: maxX, max_y: maxY };
}

export function newId(prefix: string, existing: { id: string }[]): string {
  const used = new Set(existing.map((e) => e.id));
  for (let i = 0; i < 100000; i++) {
    const candidate = `${prefix}_${i}`;
    if (!used.has(candidate)) return candidate;
  }
  throw new Error(`could not generate a unique id for prefix ${prefix}`);
}

// Find the closest wall to a point, with a maximum snap distance in
// metres. Used by the opening-add tool so doors and windows attach to
// the wall the user clicked near, not floating in space.
export function closestWall(
  floor: Floor,
  point: Point2D,
  maxDistance: number,
): { wall: Wall; projection: Point2D; distance: number } | null {
  let best: { wall: Wall; projection: Point2D; distance: number } | null = null;
  for (const wall of floor.walls) {
    const projection = projectPointToSegment(point, wall.start, wall.end);
    const dx = projection.x - point.x;
    const dy = projection.y - point.y;
    const distance = Math.sqrt(dx * dx + dy * dy);
    if (distance > maxDistance) continue;
    if (best === null || distance < best.distance) {
      best = { wall, projection, distance };
    }
  }
  return best;
}

function projectPointToSegment(point: Point2D, a: Point2D, b: Point2D): Point2D {
  const dx = b.x - a.x;
  const dy = b.y - a.y;
  const lengthSquared = dx * dx + dy * dy;
  if (lengthSquared === 0) return { x: a.x, y: a.y };
  let t = ((point.x - a.x) * dx + (point.y - a.y) * dy) / lengthSquared;
  if (t < 0) t = 0;
  else if (t > 1) t = 1;
  return { x: a.x + t * dx, y: a.y + t * dy };
}

File: apps/api/verolas_api/workflow/origin/geometry.py
"""Normalized building geometry produced by CAD parsers.

Every Origin parser (DXF, IFC, future formats) emits a `Geometry` value
of this shape. Downstream consumers (the SVG renderer in 6c.4, the
parametric grid engine in 6c.7, the AI options adapter, the 3D viewer)
read this same model. Keeping one normalized representation means each
adapter can be unit-tested without touching the raw CAD libraries.

Units: all distances are metres. Coordinates are in the floor's local
2D plane (x, y); elevation belongs to the `Floor` row, not to entities.
The parsers convert from the CAD file's native units to metres before
emitting.

Entity ids are stable per-floor. They are deterministic strings the
parser assigns ("wall_0", "wall_1", ...) so two consecutive parses of
the same file produce diff-friendly output. They are NOT meant to
survive across parses of a newly edited file.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

WallKind = Literal["exterior", "interior", "shear", "unknown"]
OpeningKind = Literal["door", "window", "opening"]
SourceFormat = Literal["dxf", "ifc"]


class Point2D(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    x: float
    y: float


class Extents(BaseModel):
    """Axis-aligned bounding box of a floor or set of entities."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    min_x: float
    min_y: float
    max_x: float
    max_y: float

    @property
    def width_m(self) -> float:
        return self.max_x - self.min_x

    @property
    def depth_m(self) -> float:
        return self.max_y - self.min_y

    @property
    def area_m2(self) -> float:
        return max(0.0, self.width_m) * max(0.0, self.depth_m)


class Wall(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    start: Point2D
    end: Point2D
    thickness_m: float = 0.20
    kind: WallKind = "unknown"

    @property
    def length_m(self) -> float:
        dx = self.end.x - self.start.x
        dy = self.end.y - self.start.y
        return float((dx * dx + dy * dy) ** 0.5)


class Opening(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    # Wall this opening pierces. May be empty when the parser cannot
    # associate the opening with a specific wall; the architectural
    # review step lets the engineer re-anchor it.
    wall_id: str = ""
    center: Point2D
    width_m: float
    kind: OpeningKind = "opening"


class Slab(BaseModel):
    """A slab outline, used to draw the floor plate."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    # Closed polygon vertices in 2D. The parser ensures the polygon is
    # closed (first == last); consumers should not assume CCW vs CW.
    polygon: list[Point2D] = Field(min_length=3)


class Column(BaseModel):
    model_config = ConfigDict(frozen=True, extra="forbid")

    id: str
    center: Point2D
    # Plan footprint of the column. (width, depth) in metres; for round
    # columns the parser sets width == depth == diameter.
    size_m: tuple[float, float] = (0.30, 0.30)


class Floor(BaseModel):
    """One storey of the building."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    key: str  # stable id, e.g. "floor_1"
    name: str  # display label, e.g. "Floor 1" or "Top Roof"
    elevation_m: float = 0.0
    extents: Extents
    walls: list[Wall] = Field(default_factory=list)
    openings: list[Opening] = Field(default_factory=list)
    slabs: list[Slab] = Field(default_factory=list)
    columns: list[Column] = Field(default_factory=list)
    # True if this floor represents a roof rather than an occupied
    # storey. The quality checker uses this for the roof-present check.
    is_roof: bool = False


class Geometry(BaseModel):
    """Top-level container for a parsed CAD file."""

    model_config = ConfigDict(extra="forbid")

    source_format: SourceFormat
    floors: list[Floor]
    # Parser warnings the caller may surface to the engineer, e.g.
    # "no layer named 'wall' found; treated all LWPOLYLINE as walls".
    parser_notes: list[str] = Field(default_factory=list)

    @property
    def floor_count(self) -> int:
        return len(self.floors)

    @property
    def wall_count(self) -> int:
        return sum(len(f.walls) for f in self.floors)

    @property
    def opening_count(self) -> int:
        return sum(len(f.openings) for f in self.floors)

    @property
    def column_count(self) -> int:
        return sum(len(f.columns) for f in self.floors)

    @property
    def slab_count(self) -> int:
        return sum(len(f.slabs) for f in self.floors)
